thunder/transforms/autodiff.py

```diff
@@ -0,0 +1,477 @@
+
+    end_time_ns = time.perf_counter_ns()
+    elapsed_time_ns = end_time_ns - start_time_ns
+    elapsed_time_millis = elapsed_time_ns // 1000000
+    trace.set_provenance(
+        thunder.core.trace.TraceProvenance(f"Grad transform pass (took {elapsed_time_millis} milliseconds)")
+    )
+    return trace
+
+
+def split_into_forward_and_backward(joint_trace):
+    """split a joint trace for forward and backward into separate ones, including recomputation (aka activation checkpointing)"""
+
+    # the joint trace will have the forward computation at the beginning and then the backward computation
+    # from how it is constructed.
+    # we split the trace:
+    # - forward symbols go into forward_part_bsyms
+    # - all symbols not in the forward go into backward_part_bsyms
+    # - for recomputation (aka activation checkpointing), we want to insert symbols going into the forward also into the
+    #   backward. but we want to do so "just in time". To this end, we gather the symbols in a dict and later
+    #   insert it when their respective outputs are needed. This is in backward_part_bsyms_recomputated
+    #   The just in time recomputation is a heuristic to save memory mimicking checkpointing: e.g. for a checkpointed
+    #   block, the forward would be recomputed just before computing the gradient.
+    # the splitting is done in the reverse order fo the bound symbols, and works out which bits are needed for the forward
+    # from there.
+    forward_part_bsyms = []
+    backward_part_bsyms = []
+    backward_part_bsyms_recomputed: dict[str, tuple[BoundSymbol, set[str]]] = {}  # name -> (bsym, name_set)
+
+    # return has a dict, it contains the forward outputs in "output" and the backward outputs in "grad_flat_args", corresponding
+    # to "flat_args"
+    return_bsym = joint_trace.bound_symbols[-1]
+    assert return_bsym.sym == prims.python_return
+    fw_output = return_bsym.args[0]["output"]
+    assert isinstance(fw_output, tuple)
+
+    grad_outs = [None for _ in fw_output]
+    output_pos = {o.name: i for i, o in enumerate(fw_output) if isinstance(o, thunder.TensorProxy)}
+
+    # the proxies we need to compute in the forward - we start with the outputs of the forward
+    forward_proxy_names = {o.name for o in thunder.core.pytree.tree_iter(fw_output) if isinstance(o, thunder.Proxy)}
+    # we also have the inputs available, so we add flat_args.
+    # for inplace, we need to update this (or have flat args be the right thing?...)
+    forward_proxy_names.update(a.name for a in return_bsym.args[0]["flat_args"] if isinstance(a, thunder.Proxy))
+
+    # We keep track of the names of proxies we recompute in the backward as those will not need to be part of the
+    # ones saved in the forward for the backward
+    backward_recomputed_proxy_names = set()
+
+    # here we go through the bound symbols to assign to the three categories, keeping track of the forward names, the backward needed
+    # and the backward available names.
```

Comment:
I think this comment is stale, as there is now only `backward_recomputed_proxy_names`.